### PR TITLE
Add vault module for SDB

### PR DESCRIPTION
**Pull request by techhat**
Ref #27020.

### CODE REVIEW

**Comment by ketzacoatl** [comment]
First read LGTM. There are a few _etcd_ strings to update. To take it for a spin, can I copy the new file somewhere into the salt extra path?

**Comment by techhat** [comment]
I don't know if there's an easy way outside of just placing it directly in the ``salt/sdb/`` directory. I'll take a look at the strings.

**Comment by ketzacoatl** [comment]
https://docs.saltstack.com/en/latest/topics/sdb/index.html does not really do much to help one understand how to use sdb. Let's say I have a jinja template, how do I make use of the vault interface through the sdb URI?

**Comment by techhat** [comment]
`salt['config.get']('sdb://myvault/secret/passwords?mypassword')`

**Comment by techhat** [comment]
From the command line, you could use something like:
```
salt-call sdb.set 'sdb://myvault/secret/salt?saltstack' super_awesome
salt-call sdb.get 'sdb://myvault/secret/salt?saltstack'
```

**Comment by ketzacoatl** [comment]
OK, that makes more sense. I have been unable to get basic tests with sdb set/get to work. Adding some error handling on missing keys / can't find vault / etc might be nice for debugging these types of user config issues.

**Comment by techhat** [comment]
@ketzacoatl, I've added a little error handling and some error messages. As for vault not being found, that's not an issue; this driver uses HTTP(S), and so does not require vault to actually be installed locally. And if you point to a URL that doesn't work, then the error handling will pick it up.

File: salt/sdb/vault.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    myvault:
+      driver: vault
+      vault.host: 127.0.0.1
+      vault.port: 8200
+      vault.scheme: http  # Default is https
+      vault.token: 012356789abcdef  # Optional
+
+The ``driver`` refers to the ``vault`` module, ``vault.host`` refers to the host
+that is hosting vault and ``vault.port`` refers to the port on that host. A
+vault token is also required. It may be set statically, as above, or as an
```
Comment (rallytime) [inline salt/sdb/vault.py]: This says the vault token is "also required" but there is a comment in code-block that says it's optional. Can you clarify this in the docs here?

File: salt/sdb/vault.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+environment variable:
+
+    export VAULT_TOKEN=0123456789abcdef
```
Comment (rallytime) [inline salt/sdb/vault.py]: Should this be in a code-block?

**Comment by rallytime** [comment]
@techhat Other than my minor doc comments, this looks good! Once that gets fixed up we can get this in.

**Comment by techhat** [comment]
@ketzacoatl I've submitted #31367 to provide better docs on using SDB. Let me know if it leaves you with any questions.

**Comment by techhat** [comment]
@rallytime, I've made clarifications. Let me know if that works for you.

**Comment by rallytime** [comment]
@techhat Looks great - thank you!

**Comment by ketzacoatl** [comment]
Great work, thanks!

**Pull request merged**